### PR TITLE
[ASTPrinter] Print the desugared constraint type following the `any` keyword.

### PR DESCRIPTION
**Pull request by hborla**
Currently, a protocol typealias with an underlying constraint type (i.e. a protocol or protocol composition) can be used as an existential type by accessing the alias through a type parameter, e.g.

```swift
protocol P1 {
  typealias A = P2
}

protocol P2 {}

func generic<T: P1>(value: (T) -> T.A) {} // okay
```

However, verifying that the underlying type of `T.A` is a constraint type is a circular computation in the general case, which is needed to use the `any` keyword. Resolving `T.A` where `T` is a type parameter needs a full generic signature to access the requirements on `T`, but a full generic signature needs the existential type verification.

This is an issue when `T.A` is written in source, but `any` is automatically printed in Swift interfaces. For now, simply print the underlying constraint type.

Resolves: rdar://93780361

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

File: test/ModuleInterface/existential-any.swift [hunk 1]
```diff
@@ -38,3 +38,11 @@ public struct S {
...
   public var q: any Q
 }
+
+
+public protocol ProtocolTypealias {
+  typealias A = P
+}
+
+// CHECK: public func dependentExistential<T>(value: (T) -> any main.P) where T : main.ProtocolTypealias
+public func dependentExistential<T: ProtocolTypealias>(value: (T) -> T.A) {}
```
Comment (tshortli) [inline test/ModuleInterface/existential-any.swift:48]: My only hesitation here is that we might want to think about whether this interferes with the ability of `-preserve-types-as-written` to help library owners work around naming ambiguities. There are already places where library owners can't control the type names that are written into the interface, though, so maybe it's not much of a regression to lack control for this kind of signature.

**Review by tshortli — APPROVED** [review]

**Review by slavapestov — APPROVED** [review]

File: test/ModuleInterface/existential-any.swift (hunk 1 above)
Comment (hborla) [reply test/ModuleInterface/existential-any.swift:48]: In any case, this isn't a long term fix; either we need to figure out how to make `any` with protocol typealiases work (seems unlikely), or eventually ban them in source.

**Comment by AnthonyLatsis** [comment]
By a cycle, do you mean something like this?

```swift
protocol P1 {
  associatedtype Assoc

  typealias A = P2
}
protocol P2 {}

func foo<T: P1>(t: T) where T.Assoc == any T.A {}
```

Sorry to bother, just trying to puzzle out some criteria for these cycles. AFAICT, they can really happen only if `any T.A` appears in a generic requirement.

**Pull request merged**